Author: Ricardo Salveti de Araujo <ricardo.salveti@canonical.com>
Title: Avoid races when sending EOS, making sure pipeline is in playing state first
Forwarded: yes
Bug: https://bugreports.qt.io/browse/QTBUG-45707
Bug-Ubuntu: https://launchpad.net/bugs/1433563

Index: qtmultimedia-opensource-src-5.4.1/src/plugins/gstreamer/mediacapture/qgstreamercapturesession.cpp
===================================================================
--- qtmultimedia-opensource-src-5.4.1.orig/src/plugins/gstreamer/mediacapture/qgstreamercapturesession.cpp
+++ qtmultimedia-opensource-src-5.4.1/src/plugins/gstreamer/mediacapture/qgstreamercapturesession.cpp
@@ -828,11 +828,11 @@ void QGstreamerCaptureSession::setState(
             if (!m_waitingForEos) {
                 m_waitingForEos = true;
                 //qDebug() << "Waiting for EOS";
+                // Unless gstreamer is in GST_STATE_PLAYING our EOS message will not be received.
+                gst_element_set_state(m_pipeline, GST_STATE_PLAYING);
                 //with live sources it's necessary to send EOS even to pipeline
                 //before going to STOPPED state
                 gst_element_send_event(m_pipeline, gst_event_new_eos());
-                // Unless gstreamer is in GST_STATE_PLAYING our EOS message will not be received.
-                gst_element_set_state(m_pipeline, GST_STATE_PLAYING);
 
                 return;
             } else {
